### ISSUE TITLE
Rewrite

### ISSUE THREAD

**Issue by Maxopoly** [Civcraft/ItemExchange#25]
Base around ItemMaps and CivModCores item name registration to allow supporting stuff like fireworks, Brewery etc. easily

Allow setting up, modifying and using shops via ItemGui

Remove need for exchange rule items with gui and external exchange storage (database?)

Add permission to limit shops via namelayer

Possibly add way to support more secure shops

**Comment by ProgrammerDan** [comment]
Note to future implementer, look at #23 for an example of how to properly encode color chars. 

Note that the color system is 'hacked' a bit by the current implementation. "invalid" codes are used; minecraft ignores those codes currently, and thus they can be used as separators and data structure indicators. However, the bukkit conversion routines won't convert the '&' or whatever char you choose to use in front of invalid pairs (`&-` for instance). Hence, the PR, where I use `replace` and the `COLOR_CHAR` directly.

**Comment by Maxopoly** [comment]
Or even better use the ChatColor enum instead. The color codes used there aren't parsed from user input or config as far as I can tell, they are hardcoded as "old color code" (for example §4) and then translated during runtime.

**Comment by ProgrammerDan** [comment]
No. I'm specifically referring to the invalid codes -- as stated, like `&-` -- which have no enum equivalent. The PR source demonstrated proper use of _both_ -- ChatColor enum where available, and the `COLOR_CHAR` solution where ChatColor considers its use invalid -- the "long-working hack".

**Comment by Maxopoly** [comment]
What would those invalid codes be needed for?

**Comment by ProgrammerDan** [comment]
They are how information is encoded into the ItemRule.

**Comment by ProgrammerDan** [comment]
From earlier comment:

> Note that the color system is 'hacked' a bit by the current implementation. "invalid" codes are used; minecraft ignores those codes currently, and thus they can be used as separators and data structure indicators

**Comment by Maxopoly** [comment]
Oh okay, I intend to remove that anyway

**Comment by Maxopoly** [comment]
in case a rewrite for this ever happens

**Comment by ProgrammerDan** [comment]
Ahhh. Yeah, that would be the "safest" route-- this is very clever approach but places its trust in minecraft not changing things up and stripping out invalid chars or some other suppression. Something using Metadata would be superior; or database tracking; but you can't discount the amazing utility of those buttons... very flexible.
